src/JuMP.jl

```diff
@@ -73,7 +73,15 @@ An abstract type that should be subtyped for users creating JuMP extensions.
 """
 abstract type AbstractModel end
 
-mutable struct Model <: AbstractModel
+"""
+    value_type(::Type{<:Union{AbstractModel,AbstractVariableRef}})
+
+Return the return type of [`value`](@ref) for variables of that model. It
+defaults to `Float64` if it is not implemented.
+"""
+value_type(::Type{<:AbstractModel}) = Float64
+
+mutable struct GenericModel{T} <: AbstractModel
```

Comment:
Could we start off by allowing a fixed list of type parameters? Like `Float32`, `Float64`, `Rational{...}`, `BigFloat`?

Reply:
Do you mean for this particular PR, or the entire sequence of PRs?

Reply:
For this PR. I mean adding a constructor that will reject type parameters outside this list.

Reply:
Would we remove it in https://github.com/jump-dev/JuMP.jl/pull/3385?

It's fair to even restrict it to only `Float64` for this PR.

Reply:
The issue is that one use case is to use solvers with https://github.com/JuliaMath/DoubleFloats.jl or https://github.com/JuliaMath/Quadmath.jl (especially for tricky SDPs). We don't want to use Julia's package extension for that :smile:

Reply:
Could we at least document which types are expected to work?
What about the method error principle when someone uses a weird type?

Reply:
How about a docstring for `Model` and a separate one for `GenericModel`. We can keep the current one for `Model`. 

I think it's hard for JuMP to make assumptions about what will work. Can we just push this onto the solver? And say something along the lines of 

```
The direct use of `GenericModel` for a value type other than `Float64` is an
advanced operation. The value type of `GenericModel` must match that expected by
the chosen optimizer. Consult the optimizers documentation for details. If not
documented, assume that the optimizer supports only `GenericModel{Float64}`.
Choosing an unsupported value type will throw an [`MOI.UnsupportedConstraint`](@ref)
or an [`MOI.UnsupportedAttribute`](@ref) error, the timing of which (during
the model construction or during a call to [`optimize!`](@ref)) depends on how
the solver is interfaced to JuMP.
```

Reply:
That sounds good for documentation. What about restricting `T<:Number` or `T<:Real`?

Reply:
`T<:Real` sounds good indeed.

Reply:
I'd prefer `T<:Number` if we want to use `Arb`s in solvers

Reply:
I talked about it offline with @kalmarek 
So `Arblib.Arf` is like `BigFloat`, `Arblib.Arb` is like `IntervalArithmetics.Interval{BigFloat}` but `Arblib.Acb` is different from `Complex{IntervalArithmetics.Interval{BigFloat}}` because `Acb` is a Euclidean ball while `Complex` would be a rectangular / norm infinity ball. Since multiplication by complex number is a rotation, it turns out that Euclidean balls work better than rectangular ball (where you loose accuracy by outer-approximating rotated rectangle by non-rotated rectangles).
The current design of the complex JuMP interface does not really support Acb at the moment so I think `Real` is fine for now.